v3: Gateway/Shard: Add Shard Manager

mavolin opened the pull request.

mavolin committed:
Gateway: Fix gateway reconnect

This commit uses the correct timeout, Gateway.ReconnectTimeout, when reconnecting. Furthermore, it adds a delay between consecutive, failed reconnects.

mavolin committed:
Gateway: Stop pacemaker when calling Gateway.CloseGracefully

mavolin committed:
API: remove unnecessary leading/trailing whitespaces

mavolin committed:
Gateway: Add Gateway.OnScalingRequired callback

mavolin committed:
Gateway: Make all user initiated user closures graceful and ensure that closures are respected during reconnects

mavolin committed:
Gateway: Fix typo

mavolin committed:
Gateway: Add Gateway.ReconnectAttempts and deprecate .ReconnectTimeout

mavolin committed:
Draft

mavolin committed:
Merge branch 'v2' into feat/ddraft/shard-manager

# Conflicts:
#	gateway/gateway.go
#	gateway/integration_test.go
#	utils/wsutil/ws.go

mavolin committed:
Gateway/Shard: Fully implement Manager

mavolin committed:
Gateway/Shard: Rename NewManagerFromXXX to NewManagerWithXXX

mavolin committed:
Gateway/Shard: Rename Manager.Apply to .ApplyError and add .Apply

mavolin committed:
Gateway/Shard: Remove remaining session.Session.Gateway references

mavolin committed:
Gateway/Shard: Attempt to Open, Close, Pause and UpdateStatus all other gateways before returning with an error

chanbakjsd commented on gateway/shard/manager.go:
```diff
@@ -0,0 +1,327 @@
+
+	gatewayURL := botData.URL + "?" + param.Encode()
+
+	m := newIdentifiedManager(gatewayURL, id, botData.Shards, GenerateShardIDs(botData.Shards)...)
```
This code seems to be duplicated with the code below.

chanbakjsd commented on gateway/shard/manager.go:
```diff
@@ -0,0 +1,327 @@
+	gateways := make([]*gateway.Gateway, len(shardIDs))
+
+	for i, shardID := range shardIDs {
+		id.SetShard(shardID, totalShards)
+		idCp := *id
+
+		gateways[i] = gateway.NewCustomIdentifiedGateway(gatewayURL, &idCp)
```
`Shard` seems to be a pointer to an array internally (`*[2]int`) so copying like this will only copy the pointer and not the data itself.
This might cause some race condition.

chanbakjsd commented on gateway/shard/manager.go:
```diff
@@ -0,0 +1,327 @@
+	return m, nil
+}
+
+// NewIdentifiedManager creates a new Manager using the passed url and the
+// passed gateway.Identifier. The shard information stored on the passed
+// identifier will be ignored. Instead totalShards and shardIDs will be used.
```
It might be worth noting that `NewIdentifiedManager` does not handle OnShardingRequired in the docs as I think the Manager's docs for OnShardingRequired seems to imply that OnShardingRequired is always present somewhat.

mavolin replied:
`OnShardingRequired` is actually never present unless using `NewManager`. So I'll try use better wording for both docs.

chanbakjsd commented on gateway/shard/manager.go:
```diff
@@ -0,0 +1,327 @@
+// ApplyError is the same as Apply, but the iterator function returns an error.
+// If such an error occurs, the error will be returned wrapped in an *Error.
+//
+// If all is set to true, ApplyError will apply the passed function to all
+// gateways. If a single error occurs, it will be returned as an *Error, if
+// multiple errors occur then they will be returned as *MultiError.
```
`all` is unimplemented.

chanbakjsd commented on gateway/shard/manager.go:
```diff
@@ -0,0 +1,327 @@
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	for i, g := range m.gateways {
+		if err := g.Open(); err != nil {
+			errs = append(errs, &Error{
+				ShardID: shardID(g),
+				Source:  err,
+			})
+
+			for _, g := range m.gateways[:i] {
+				if err := g.Close(); err != nil {
+					errs = append(errs, &Error{
+						ShardID: shardID(g),
+						Source:  err,
+					})
+				}
+			}
+
+			if len(errs) == 1 {
+				return errs[0]
+			}
+
+			return errs
+		}
+	}
```
This code can use ApplyError instead.

mavolin replied:
It actually can't, because unlike the other methods like `Close`, `Open` attempts to close all previously opened gateways if it encounters an error.
This was done, because in most cases if you `Open` your session/state during startup, and encounter an error, you would most likely exit the program.

chanbakjsd replied:
You can probably just loop over all gateways again to close them if `ApplyError` returned an error unless I'm missing something.

mavolin replied:
No, because in an error scenario we would need the index of the gateway that returned an error, so we only close those gateways (i.e `m.gateways[:i]`, see line 248) that were previously opened.

chanbakjsd replied:
`ShardID` can provide that information. If the operation of searching based on ShardID is too expensive, it might be a sign that `Error` could use an "internal ID" to be more useful.

mavolin replied:
Sure, I can do that. Computational expense is negligible since: a) open is ever only performed once, b) network is the true bottleneck here, and c) there will never be a large number of shards stored in a manager.

chanbakjsd commented on gateway/shard/manager.go:
```diff
@@ -0,0 +1,327 @@
+	return m.ApplyError(func(g *gateway.Gateway) error { return g.UpdateStatus(d) }, true)
+}
+
+func (m *Manager) RequestGuildMembers(d gateway.RequestGuildMembersData) error {
```
Exported function does not have documentation.

chanbakjsd replied:
I don't think copy pasting the documentation is the right move here. Especially considering it's going to get squashed into one row on pkg.go.dev.

mavolin replied:
All docs for API related methods are copied from the official docs. The docs are usually the most extensive information we have, so if I were to write docs myself, I would basically just be rewording the docs found on the official API doc page.

chanbakjsd commented on internal/moreatomic/bool.go:
```diff
@@ -17,3 +17,12 @@ func (b *Bool) Set(val bool) {
+		oldN = 1
+	}
+
+	return atomic.CompareAndSwapUint32(&b.val, oldN, oldN+1%2)
```
```suggestion
	return atomic.CompareAndSwapUint32(&b.val, oldN, (oldN+1) % 2)
```
The remainder operation is done before the add operation similar to how multiplication is done before add.

chanbakjsd replied:
```suggestion
	return atomic.CompareAndSwapUint32(&b.val, oldN, (oldN+1) % 2)
```
The remainder operation is done before the addition operation similar to multiplication and addition.

chanbakjsd replied:
```suggestion
	return atomic.CompareAndSwapUint32(&b.val, oldN, (oldN+1) % 2)
```
The remainder operation is done before the addition operation similar to multiplication and addition.

chanbakjsd commented on gateway/shard/shard.go:
```diff
@@ -0,0 +1,81 @@
+	"strings"
+)
+
+// GenerateShardIDs generates an array of ints of 0..(total-1).
```
```suggestion
// GenerateShardIDs generates a slice of ints of [0, total).
```
Array means a different thing in Go.

chanbakjsd replied:
```suggestion
// GenerateShardIDs generates a slice of integers of [0, total).
```
Arrays mean a different thing in Go.

chanbakjsd commented on gateway/shard/shard.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+func (err *Error) Error() string {
+	panic("implement me")
```
Please implement this.

chanbakjsd replied:
Please implement this.

chanbakjsd commented on gateway/shard/shard.go:
```diff
@@ -0,0 +1,81 @@
+type Error struct {
+	// ShardID is the id of the shard that returned the error.
+	ShardID int
+	// Source is the error itself
```
Inconsistent full stop.

chanbakjsd commented on gateway/shard/shard.go:
```diff
@@ -0,0 +1,81 @@
+
+func (errs MultiError) Error() string {
+	var b strings.Builder
+	n := 24 + 2*len(errs)
```
24 feels too magical. Maybe make `"multiple errors occurred:"` a const string and do `len` here instead.

chanbakjsd replied:
24 seems like too magical of a number. Consider making the string an unexported const and `len` it instead.

chanbakjsd commented on session/session.go:
```diff
@@ -47,19 +49,27 @@ func NewWithIntents(token string, intents ...gateway.Intents) (*Session, error)
 	}
 
-	return NewWithGateway(g), nil
+	return NewWithGateways(g), nil
```
Is it really ideal to only start one Gateway when Intents are specified?

chanbakjsd commented on session/session.go:
```diff
@@ -47,19 +49,27 @@ func NewWithIntents(token string, intents ...gateway.Intents) (*Session, error)
 	}
 
-	return NewWithGateway(g), nil
+	return NewWithShardManager(m), err
```
`err` is guaranteed to be `nil` so just return `nil` here.

chanbakjsd commented on state/state.go:
```diff
@@ -119,7 +120,24 @@ func NewWithStore(token string, cabinet store.Cabinet) (*State, error) {
+
+		m, err := shard.NewManager(token)
+		if err != nil {
+			return nil
```
Return an error in this function too instead of ignoring the error.

mavolin replied:
The idea here is, that if such an error occurs, all connections are closed. If the user wants to handle that error, e.g. log it, implement a more complex retry mechanism, they would have to provide a custom implementation.
I did this, because adding another logging/capture function that will ever just receive a single error seems way to bloated, especially since the above solution is just as elegant.

Edit: If it wasn't apparent, the connections are closed if `nil` is returned.

chanbakjsd replied:
Ah right. I was reading this as part of `newWithAutoRescale` instead of in `state.ShardManager.OnShardingRequired`. Feel free to resolve this.

chanbakjsd commented on state/state.go:
```diff
@@ -666,14 +694,16 @@ func (s *State) Messages(channelID discord.ChannelID) ([]discord.Message, error)
 // will look for the presence in all cached guilds.
-func (s *State) Presence(gID discord.GuildID, uID discord.UserID) (*gateway.Presence, error) {
-	if !s.Gateway.HasIntents(gateway.IntentGuildPresences) {
+func (s *State) Presence(
+	guildID discord.GuildID, userID discord.UserID) (*gateway.Presence, error) {
+
```
Unnecessary indent change?

mavolin replied:
The function header was longer than the 100 character limit, so I wrapped it.

chanbakjsd commented on state/state.go:
```diff
@@ -683,21 +713,22 @@ func (s *State) Presence(gID discord.GuildID, uID discord.UserID) (*gateway.Pres
 
-func (s *State) Role(guildID discord.GuildID, roleID discord.RoleID) (target *discord.Role, err error) {
-	if s.Gateway.HasIntents(gateway.IntentGuilds) {
+func (s *State) Role(guildID discord.GuildID, roleID discord.RoleID) (target *discord.Role,
+	err error) {
```
More unnecessary indent change?

mavolin replied:
The same here with the 100 character limit, but in this case it seems my IDE's formatter wrapped this incorrectly, so fixing this.

chanbakjsd commented on state/state.go:
```diff
@@ -753,17 +784,19 @@ func (s *State) Roles(guildID discord.GuildID) ([]discord.Role, error) {
-	m, err = s.Session.Member(gID, uID)
-	if err == nil && s.Gateway.HasIntents(gateway.IntentGuildMembers) {
-		err = s.Cabinet.MemberSet(gID, *m)
+func (s *State) fetchMember(
+	guildID discord.GuildID, userID discord.UserID) (m *discord.Member, err error) {
```
Why is this pushed to the next line?

mavolin replied:
Again, 100 character line limit

chanbakjsd commented on state/state.go:
```diff
@@ -772,12 +805,14 @@ func (s *State) fetchMember(gID discord.GuildID, uID discord.UserID) (m *discord
 func (s *State) tracksMessage(m *discord.Message) bool {
-	return (m.GuildID.IsValid() && s.Gateway.HasIntents(gateway.IntentGuildMessages)) ||
-		(!m.GuildID.IsValid() && s.Gateway.HasIntents(gateway.IntentDirectMessages))
+	return (m.GuildID.IsValid() &&
+		s.ShardManager.FromGuildID(m.GuildID).HasIntents(gateway.IntentGuildMessages)) ||
+		(!m.GuildID.IsValid() &&
+			s.ShardManager.FromGuildID(m.GuildID).HasIntents(gateway.IntentDirectMessages))
```
Might be worth lifting `s.ShardManager.FromGuildID(m.GuildID)` into a variable to maintain code similar to the old code.

chanbakjsd commented on state/state.go:
```diff
@@ -772,12 +805,14 @@ func (s *State) fetchMember(gID discord.GuildID, uID discord.UserID) (m *discord
 func (s *State) tracksChannel(c *discord.Channel) bool {
-	return (c.GuildID.IsValid() && s.Gateway.HasIntents(gateway.IntentGuilds)) ||
-		!c.GuildID.IsValid()
+	return (c.GuildID.IsValid() && s.ShardManager.FromGuildID(c.GuildID).
+		HasIntents(gateway.IntentGuilds)) || !c.GuildID.IsValid()
```
Weird indentation.

mavolin replied:
This is the indention gofmt uses, what should I change? I could put the function call on the same line, but that would require an extra line.

chanbakjsd replied:
I personally think another line doesn't hurt and will aid in readability instead.

chanbakjsd reviewed (changes requested):
Nice work. This PR adds shard manager to arikawa and moves `Session` (and by extension, `State` and `Bot` too) on top of it.
There are a few problems that needs to be addressed but otherwise this PR looks great!

mavolin committed:
Fix style issues and simplify code

mavolin committed:
Gateway/Shard: Clarify behavior on OnScalingRequired = nil or return nil

chanbakjsd commented on gateway/shard/manager.go:
```diff
@@ -206,16 +194,30 @@ func (m *Manager) ApplyError(f func(g *gateway.Gateway) error, all bool) error {
-	return nil
+	if len(errs) == 0 {
+		return nil
+	} else if len(errs) == 1 {
```
Unnecessary `else` here. In fact, consider making it a `switch` on `len(errs)` instead.

mavolin replied:
Why is the else unnecessary? If we just encounter a single error, we should return just that one instead of the MultiError. Also we only have two cases, so why a switch?

chanbakjsd replied:
The `else` is not required because the `if` contains an early return. A `switch` can contain `case 0:`, `case 1:` and `default:`.

chanbakjsd reviewed (changes requested):
The changes made had a few minor issues in it.

mavolin committed:
Gateway/Shard: Code cleanup

mavolin committed:
State: Fix improper reset of State when OnScalingRequired is called

mavolin committed:
Gateway/Shard: Rename Manager.OnScalingRequired to .Rescale and replace 'Scaling Required' with `Sharding Required` in docs

There is no error code 'Scaling Required', but only 'Sharding Required'. Therefore all incorrect references are removed. Additionally Manager.OnScalingRequired is renamed to a more fitting .Rescale.

mavolin committed:
Merge branch 'v3' into feat/ddraft/shard-manager

# Conflicts:
#	bot/ctx.go
#	internal/moreatomic/guildid_set.go
#	session/session.go

mavolin committed:
State: Use State.Reset instead of State.Cabinet.Reset on ready event

mavolin committed:
store: fix incorrect field reference after v3 merge

diamondburned commented on gateway/shard/manager.go:
```diff
@@ -0,0 +1,375 @@
+			return
+		}
+
+		m.mutex.Lock()
```
This mutex should probably be moved to be acquired after `Rescale()`, just to prevent any potential deadlocks.

diamondburned commented on voice/voice.go:
```diff
@@ -4,11 +4,16 @@
-func AddIntents(gw *gateway.Gateway) {
-	gw.AddIntents(gateway.IntentGuilds)
-	gw.AddIntents(gateway.IntentGuildVoiceStates)
+func AddIntents(m *shard.Manager) {
```
I think this should be split into a version for `Gateway` and another for the shard manager, just to make things just as easy as before for people implementing their own shard manager.

mavolin replied:
I think instead of having two methods, one for gateway and one for shard.Manager, it might be simpler to just have an expose `Intents` constant, that users add themselves.

diamondburned replied:
That works as well.

diamondburned commented on state/state.go:
```diff
@@ -119,7 +121,24 @@ func NewWithStore(token string, cabinet store.Cabinet) (*State, error) {
+func newWithAutoRescale(s *session.Session, cabinet store.Cabinet) *State {
+	state := NewFromSession(s, cabinet)
+	state.ShardManager.Rescale = func() *shard.Manager {
+		state.Reset()
```
`Reset` is not only called here but also on `Ready` events, which would cause the state to be incomplete when the shards are starting up. A way to fix this would probably be to have a `noResetOnReady bool` that prevents the handlers from resetting the state there, so we can do it here.

This call should also be moved to after the `NewManager` call.

mavolin replied:
If we expect people to implement their own shard manager, we should probably expose `noResetOnReady`.

diamondburned commented on state/state.go:
```diff
@@ -6,6 +6,8 @@ import (
 	"sync"
 
+	"github.com/diamondburned/arikawa/v2/gateway/shard"
+
```
Random new line?

mavolin committed:
shard: Code cleanup

mavolin committed:
Merge branch 'v3' into feat/ddraft/shard-manager

# Conflicts:
#	state/state.go
